soc/intel/intel_adsp/tools/cavstool.py

```diff
@@ -461,6 +495,130 @@ def load_firmware(fw_file):
+        fw_bytes = fw_bytes[sz:len(fw_bytes)]
+
+    # This actually means "enable access to BAR4 registers"!
+    hda.PPCTL |= (1 << 30) # GPROCEN, "global processing enable"
+
+    log.info("Resetting HDA device")
+    hda.GCTL = 0
+    while hda.GCTL & 1: pass
+    hda.GCTL = 1
+    while not hda.GCTL & 1: pass
+
+    log.info("Turning of DSP subsystem")
+    dsp.HFDSSCS &= ~(1 << 16) # clear SPA bit
+    time.sleep(0.002)
+    # wait for CPA bit clear
+    while dsp.HFDSSCS & (1 << 24):
+        log.info("Waiting for DSP subsystem power off")
+        time.sleep(0.1)
+
+    log.info("Turning on DSP subsystem")
+    dsp.HFDSSCS |= (1 << 16) # set SPA bit
+    time.sleep(0.002) # needed as the CPA bit may be unstable
+    # wait for CPA bit
+    while not dsp.HFDSSCS & (1 << 24):
+        log.info("Waiting for DSP subsystem power on")
+        time.sleep(0.1)
+
+    log.info("Turning on Domain0")
+    dsp.HFPWRCTL |= 0x1 # set SPA bit
+    time.sleep(0.002) # needed as the CPA bit may be unstable
+    # wait for CPA bit
+    while not dsp.HFPWRSTS & 0x1:
+        log.info("Waiting for DSP domain0 power on")
+        time.sleep(0.1)
+
+    log.info("Turning off Primary Core")
+    dsp.DSP2CXCTL_PRIMARY &= ~(0x1) # clear SPA
+    time.sleep(0.002) # wait for CPA settlement
+    while dsp.DSP2CXCTL_PRIMARY & (1 << 8):
+        log.info("Waiting for DSP primary core power off")
+        time.sleep(0.1)
+
+
+    log.info(f"Configuring HDA stream {hda_ostream_id} to transfer firmware image")
+    (buf_list_addr, num_bufs) = setup_dma_mem(fw_bytes)
+    sd.CTL = 1
+    while (sd.CTL & 1) == 0: pass
+    sd.CTL = 0
+    while (sd.CTL & 1) == 1: pass
+    sd.CTL |= (1 << 20) # Set stream ID to anything non-zero
+    sd.BDPU = (buf_list_addr >> 32) & 0xffffffff
+    sd.BDPL = buf_list_addr & 0xffffffff
+    sd.CBL = len(fw_bytes)
+    sd.LVI = num_bufs - 1
+    hda.PPCTL |= (1 << hda_ostream_id)
+
+    # SPIB ("Software Position In Buffer") is an Intel HDA extension
+    # that puts a transfer boundary into the stream beyond which the
+    # other side will not read.  The ROM wants to poll on a "buffer
+    # full" bit on the other side that only works with this enabled.
+    hda.SPBFCTL |= (1 << hda_ostream_id)
+    hda.SD_SPIB = len(fw_bytes)
+
+
+    # Send the DSP an IPC message to tell the device how to boot.
+    # Note: with cAVS 1.8+ the ROM receives the stream argument as an
+    # index within the array of output streams (and we always use the
+    # first one by construction).  But with 1.5 it's the HDA index,
+    # and depends on the number of input streams on the device.
+    stream_idx = 0
+    ipcval = (  (1 << 31)            # BUSY bit
+                | (0x01 << 24)       # type = PURGE_FW
+                | (1 << 14)          # purge_fw = 1
+                | (stream_idx << 9)) # dma_id
+    log.info(f"Sending IPC command, HFIPCXIDR = 0x{ipcval:x}")
+    dsp.HFIPCXIDR = ipcval
+
+
```

Comment:
Here too.  Maybe these are intended to break up sections by theme or whatever?  If so maybe a comment would be better?  Not a big deal.